### DISCUSSION TITLE
Implement chats in a MAUI app with ChatGPT valid for a session

### DISCUSSION THREAD

**Discussion by erossini** [RageAgainstThePixel/OpenAI-DotNet#373] [Q&A] [answered]
In my `.NET 8 MAUI`, I like to implement a chat with different ChatGPT users about different topics. 

I saw your examples but none of them explains how to create more messages in the same chat with the same `conversationId`. I like to get a `json` to manage the answer from ChatGpt better. For example:
```csharp
    using var api = new OpenAIClient();
    var messages = new List<Message>
    {
        new Message(Role.System, "You are a helpful assistant designed to output JSON."),
        new Message(Role.User, "Who won the world series in 2020?"),
    };
    var chatRequest = new ChatRequest(messages, Model.GPT4o, 
                                      responseFormat: ChatResponseFormat.Json);
    var response = await api.ChatEndpoint.GetCompletionAsync(chatRequest);
    
    foreach (var choice in response.Choices)
    {
        Console.WriteLine($"[{choice.Index}] {choice.Message.Role}: {choice} " + 
                          $"| Finish Reason: {choice.FinishReason}");
    }
    
    response.GetUsage();
```
In this example, there is no interaction with a user. `chatRequest` doesn't contain a `conversationId` and I can't find how to pass this value as a context for ChatGPT if I want to add a new message from the user. 

Also, in the app, the user can leave the chat page and then open it later but the conversation should continue. Do you have any example for that?

Thank you in advance,
Enrico

**Comment by StephenHodgson** [comment, marked as answer]
Hi @erossini 

There is a [Conversation](https://github.com/RageAgainstThePixel/OpenAI-DotNet/blob/main/OpenAI-DotNet/Chat/Conversation.cs) class you can use for exactly this purpose. 

I use it like so:
```csharp
    using var api = new OpenAIClient();
    var conversation = new Conversation(
        [
            new(Role.System, "You are a helpful assistant designed to output JSON."),
            new(Role.User, "Who won the world series in 2020?")
        ]);
        
    var chatRequest = new ChatRequest(conversation.Messages, Model.GPT4o, 
                                      responseFormat: ChatResponseFormat.Json);
    var response = await api.ChatEndpoint.GetCompletionAsync(chatRequest);
    conversation.AppendMessage(response.FirstChoice.Message);
```
You should be able to serialize this object to json if you wish to store a user session and return to it later :)

also, appending the message also works for tool calls and other function/roles.

you will need your own system of remembering and saving sessions though.

**Reply by erossini** [reply]
Thank you for this code and in advance for my next question 😁. That means I must have a global `conversation` object in my application and use it when it needs it, right?

So, the next question is: what if I want to move this logic in an API? How can I keep the conversation going?

**Reply by StephenHodgson** [reply]
not global, probably you can create one per session.

the conversation can be easily converted to json, and you'd store that locally or in our database somewhere.

you'll need to keep track of the session info yourself based on any info you'd like (userid, time, etc)

to continue the conversation just deserialize the conversation and append any new questions/messages to the conversation object and pass the whole thing back in the chat request. Be aware there is a context limit. (That's why most people have moved to assistants api since it handles most of this for you by default with threads, etc.)

**Reply by erossini** [reply]
Thank you. So, it is convenient to move to `Assistant`.

**Reply by StephenHodgson** [reply]
yes, prob more aligned with your needs. You'll be able to get a threadId. Most of the features are the same, just the API is slightly different.

**Comment by erossini** [comment]
I'm trying to create the `Assistant` and adding messages after the creation. 

```csharp
    var request = new CreateAssistantRequest(Model.GPT4o);
    assistant = await client.AssistantsEndpoint.CreateAssistantAsync(
        new CreateAssistantRequest(
            name: "English Teacher",
            instructions: "You are a personal English Teacher.",
            model: Model.GPT4o));

    var messages = new List<Message> { "Good morning! How is the weather?" };
    var threadRequest = new CreateThreadRequest(messages);

    chats = await assistant.CreateThreadAndRunAsync(threadRequest);
```

Now, I can add messages only in `messages` or `threadRequest`. That means I have to re-call every time the `CreateThreadAndRunAsync` in order to update the chat with the user?

**Reply by StephenHodgson** [reply]
your `messages` variable is your local copy. You'll need to get the latest messages from the thread.
